src/utils/list_cpu_features.c

```diff
@@ -30,7 +36,176 @@
+// Design principles
+// -----------------
+// We build a tree structure containing all the data to be displayed.
+// Then depending on the output type (text or json) we walk the tree and display
+// the data accordingly.
+
+// We use a bump allocator to allocate strings and nodes of the tree,
+// Memory is not intented to be reclaimed.
+typedef struct {
+  char* ptr;
+  size_t size;
+} BumpAllocator;
+
+// Allocate a buffer of size `size`.
+static BumpAllocator BA_Create(size_t size) {
+  char* const ptr = (char*)malloc(size);
+  BumpAllocator BA;
+  if (ptr) BA = (BumpAllocator){.ptr = ptr, .size = size};
+  return BA;
+}
+
+// Update the available memory left in the BumpAllocator.
+static void* BA_Bump(BumpAllocator* BA, size_t size) {
+  assert(BA->size >= size);
+  void* ptr = BA->ptr;
+  BA->size -= size;
+  BA->ptr += size;
+  return ptr;
+}
+
+// The type of the nodes in the tree.
+typedef enum {
+  TNT_INVALID,
+  TNT_INT,
+  TNT_MAP,
+  TNT_MAP_ENTRY,
+  TNT_ARRAY,
+  TNT_ARRAY_ELEMENT,
+  TNT_STRING,
+} TreeValueType;
+
+// The node in the tree.
+typedef struct TreeValue {
+  TreeValueType type;
+  unsigned integer;
+  const char* string;
+  struct TreeValue* value;
+  struct TreeValue* next;
+} TreeValue;
+
+// Allocates a node inside a BumpAllocator.
+static TreeValue* BA_TreeValue(BumpAllocator* BA, TreeValueType type) {
+  TreeValue* TV = (TreeValue*)BA_Bump(BA, sizeof(TreeValue));
+  assert(TV);
+  TV->type = type;
+  return TV;
+}
+
+// Allocates an integer node inside a BumpAllocator.
+static TreeValue* CreateInt(BumpAllocator* BA, int value) {
+  TreeValue* TV = BA_TreeValue(BA, TNT_INT);
+  TV->integer = value;
+  return TV;
+}
+
+// Allocates a string node inside a BumpAllocator.
+// `value` must outlive the tree.
+static TreeValue* CreateConstantString(BumpAllocator* BA, const char* value) {
+  TreeValue* TV = BA_TreeValue(BA, TNT_STRING);
+  TV->string = value;
+  return TV;
+}
+
+// Allocates a map node inside a BumpAllocator.
+static TreeValue* CreateMap(BumpAllocator* BA) {
+  TreeValue* TV = BA_TreeValue(BA, TNT_MAP);
+  TV->next = NULL;
+  return TV;
+}
+
+// Allocates an array node inside a BumpAllocator.
+static TreeValue* CreateArray(BumpAllocator* BA) {
+  TreeValue* TV = BA_TreeValue(BA, TNT_ARRAY);
+  TV->next = NULL;
+  return TV;
+}
+
+// Allocates a formatted string inside a BumpAllocator.
+static TreeValue* CreatePrintfString(BumpAllocator* BA, const char* format,
+                                     ...) {
+  va_list arglist;
+  va_start(arglist, format);
+  char* const ptr = BA->ptr;
+  const int written = vsnprintf(ptr, BA->size, format, arglist);
+  va_end(arglist);
+  if (written < 0 || written >= BA->size) return NULL;
+  return CreateConstantString(BA, (char*)BA_Bump(BA, written));
+}
+
+static TreeValue* CreateString(BumpAllocator* BA, const char* value) {
+  return CreatePrintfString(BA, "%s", value);
+}
+
+// Allocates a map entry node inside a BumpAllocator.
```

Comment:
nit: technically, this add an entry to |map|. The BumpAllocator is an implementation detail, and it would probably makes more sense to put it as the last parameter.